go/vt/vttablet/tabletserver/throttle/throttler.go

```diff
@@ -394,105 +506,115 @@ func (throttler *Throttler) isDormant() bool {
 // run the probes, colelct metrics, refresh inventory, etc.
 func (throttler *Throttler) Operate(ctx context.Context) {
 
+	tickers := [](*timer.SuspendableTicker){}
 	addTicker := func(d time.Duration) *timer.SuspendableTicker {
-		throttler.initMutex.Lock()
-		defer throttler.initMutex.Unlock()
-
-		t := timer.NewSuspendableTicker(d, true)
-		throttler.tickers = append(throttler.tickers, t)
+		t := timer.NewSuspendableTicker(d, false)
+		tickers = append(tickers, t)
 		return t
 	}
-
 	leaderCheckTicker := addTicker(leaderCheckInterval)
 	mysqlCollectTicker := addTicker(mysqlCollectInterval)
 	mysqlDormantCollectTicker := addTicker(mysqlDormantCollectInterval)
 	mysqlRefreshTicker := addTicker(mysqlRefreshInterval)
 	mysqlAggregateTicker := addTicker(mysqlAggregateInterval)
 	throttledAppsTicker := addTicker(throttledAppsSnapshotInterval)
 
-	for {
-		select {
-		case <-leaderCheckTicker.C:
-			{
-				func() {
-					throttler.initMutex.Lock()
-					defer throttler.initMutex.Unlock()
+	go func() {
+		defer log.Infof("Throttler: Operate terminated, tickers stopped")
+		for _, t := range tickers {
+			defer t.Stop()
+			// since we just started the tickers now, speed up the ticks by forcng an immediate tick
+			go t.TickNow()
+		}
 
-					// sparse
-					shouldBeLeader := int64(0)
-					if atomic.LoadInt64(&throttler.isOpen) > 0 {
-						if throttler.tabletTypeFunc() == topodatapb.TabletType_PRIMARY {
-							shouldBeLeader = 1
+		for {
+			select {
+			case <-ctx.Done():
+				return
+			case <-leaderCheckTicker.C:
+				{
+					func() {
+						throttler.initMutex.Lock()
+						defer throttler.initMutex.Unlock()
+
+						// sparse
+						shouldBeLeader := int64(0)
+						if atomic.LoadInt64(&throttler.isOpen) > 0 {
+							if throttler.tabletTypeFunc() == topodatapb.TabletType_PRIMARY {
+								shouldBeLeader = 1
+							}
```

Comment:
Feels odd and potentially problematic that the throttler maintains leader data that at is least in theory independent of topo state. Assuming that's a Freno holdover? I would expect these transitions to be limited to Open() and Close(), but maybe not?

Reply:
You are right. And this is kind of a leftover from where we were, where the throttler was statically configured. I should edit this.

Reply:
Ah, no, actually! This is intentional and required. So the story is that each tablet runs a throttler. Each throttler at the very least runs "self checks" isolated from other tablets.. But in addition, the `PRIMARY` tablet also runs shard checks, where it collects data from other tablets. So the throttlers are all `Open()`, and still need to figure out, on top of being open, whether one is `PRIMARY` or not.